e2e/e2e_test.go

```diff
@@ -169,9 +398,90 @@ func (c *ClientInfo) deletePod(pod *core.Pod) error {
 		return err
 	}
 
-	if err := WaitForPodToDisappear(c.Client, pod.GetNamespace(), pod.GetName(), deleteTimeout); err != nil {
+	if err := WaitForPodToDisappear(c.Client, pod.GetNamespace(), pod.GetName(), podDeleteTimeout); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *ClientInfo) provisionReplicaSet(replicaCount int32, label, annotations map[string]string) (*apps.ReplicaSet, error) {
+	replicaSet, err := c.Client.AppsV1().ReplicaSets(testNamespace).Create(
+		context.Background(),
+		replicaSetObject(replicaCount, label, annotations),
+		metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := WaitForPodBySelector(c.Client, testNamespace, wbLabelEqual, rsCreateTimeout); err != nil {
+		return nil, err
+	}
+
+	replicaSet, err = c.Client.AppsV1().ReplicaSets(testNamespace).Get(context.Background(), replicaSet.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return replicaSet, nil
+}
+
+func (c *ClientInfo) updateReplicaSet(replicaSet *apps.ReplicaSet) (*apps.ReplicaSet, error) {
+	replicaSet, err := c.Client.AppsV1().ReplicaSets(testNamespace).Update(context.Background(), replicaSet, metav1.UpdateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return replicaSet, nil
+}
+
+func (c *ClientInfo) deleteReplicaSet(replicaSet *apps.ReplicaSet) error {
+	const rsDeleteTimeout = 2 * rsCreateTimeout
+	if err := c.Client.AppsV1().ReplicaSets(testNamespace).Delete(context.Background(), replicaSet.Name, metav1.DeleteOptions{}); err != nil {
+		return err
+	}
+
+	if err := WaitForReplicaSetToDisappear(c.Client, replicaSet.GetNamespace(), replicaSet.GetName(), rsDeleteTimeout); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v envVars) maxReplicas(allPods []core.Pod) int32 {
+	const maxPodsPerNode = 110
+	return int32(
+		(v.numComputeNodes*maxPodsPerNode - (len(allPods))) * v.fillPercentCapacity / 100)
+}
+
+// Waits for all replicas to be fully removed from replicaset, and checks that there are 0 ip pool allocations
+func checkZeroIPPoolAllocationsAndReplicas(clientInfo *ClientInfo, k8sIPAM *kubeClient.KubernetesIPAM) error {
+	const (
+		emptyReplicaSet = 0
+		rsSteadyTimeout = 2 * rsCreateTimeout
+	)
+	var err error
+
+	// Set to 0 replicas
+	replicaSet, err := clientInfo.updateReplicaSet(replicaSetObject(
+		emptyReplicaSet,
+		podTierLabel(rsName),
+		podNetworkSelectionElements(testNetworkName),
+	))
+	if err != nil {
 		return err
 	}
+
+	if err = WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, wbLabelEqual, replicaSet, rsSteadyTimeout); err != nil {
+		return err
+	}
+
+	// Check for 0 IP pool allocations
+	ipPool, err := k8sIPAM.GetIPPool(context.Background(), ipPoolName)
+	if err != nil {
+		return err
+	}
+
+	if len(ipPool.Allocations()) != 0 {
+		return fmt.Errorf("Expected 0 IP pool allocations, got %d instead", len(ipPool.Allocations()))
```

Comment:
I'm seeing this fail in the [e2e tests](https://github.com/k8snetworkplumbingwg/whereabouts/runs/5693681338?check_suite_focus=true#step:7:47); I think it would help somehow to know in which iteration did the test fail.

How about letting the called indicate that ? Something like (in line [152](https://github.com/k8snetworkplumbingwg/whereabouts/pull/181/files#diff-e91d29c6d4c58503d9c2d72e2b04f1b70901a613e23008d1d1cfe8beff7038d4R152)):
```golang
Expect(checkZeroIPPoolAllocationsAndReplicas(clientInfo, k8sIPAM)).To(Succeed(), fmt.Sprintf("Stale IP allocations in iteration: %d", i))
```

Reply:
Miguel, for the record, I somehow missed a lot of these comments from 10 days ago. Apologies! I'm catching up on these, but for this comment I think (for now) the problem is solved. I created a WaitForZeroIPPoolAllocations() function that gives kubernetes adequate time to clean up. I'm going to leave this comment unresolved, because it's a great idea in the event the problem resurfaces.